site/site.go

```diff
@@ -672,3 +720,67 @@ func RenderStaticErrorPage(rw http.ResponseWriter, r *http.Request, data ErrorPa
 		return
 	}
 }
+
+type binHashCache struct {
+	binFS http.FileSystem
+
+	hashes map[string]string
+	mut    *sync.RWMutex
+	sf     *singleflight.Group
+	sem    chan struct{}
+}
+
+func newBinHashCache(binFS http.FileSystem, binHashes map[string]string) *binHashCache {
+	b := &binHashCache{
+		binFS:  binFS,
+		hashes: make(map[string]string, len(binHashes)),
+		mut:    &sync.RWMutex{},
+		sf:     &singleflight.Group{},
+		sem:    make(chan struct{}, 4),
+	}
+	// Make a copy since we're gonna be mutating it.
+	for k, v := range binHashes {
+		b.hashes[k] = v
+	}
+
+	return b
+}
+
+func (b *binHashCache) getHash(name string) (string, error) {
+	b.mut.RLock()
+	hash, ok := b.hashes[name]
+	b.mut.RUnlock()
+	if ok {
+		return hash, nil
+	}
+
+	// Avoid DOS by using a pool, and only doing work once per file.
+	v, err, _ := b.sf.Do(name, func() (interface{}, error) {
```

Comment:
Did I understand the logic correctly that the purpose of this is to be able to compute hashes of served binaries that were not in `coder.sha1`, e.g. when it's missing?

Reply:
Yep, basically. This also serves the case where we're serving directly from the site FS because the binaries weren't compressed.

Reply:
Ok, perfect. Thanks for confirming!